src/components/utils/include/utils/timer.h

```diff
@@ -90,41 +106,106 @@ class Timer {
   Milliseconds GetTimeout() const;
 
  private:
-  const std::string name_;
-  const TimerTask* task_;
-  bool repeatable_;
-  uint32_t timeout_ms_;
-  bool is_running_;
-  timer_t timer_;
-  mutable sync_primitives::Lock lock_;
-  sync_primitives::Lock task_lock_;
-
   /**
    * @brief method called from friend handler_wrapper and call run() from task.
    */
-  void OnTimeout();
+  void OnTimeout() const;
 
   /**
-   * @brief method for setting correct timeout.
-   * @param timeout - if it`s value = 0, timeout will be setted to 1
-   * There would be no way to stop thread if timeout in lopper will be 0
-   * and if we puts to timer_create zero timeout then we get sys error(22)
+   * @brief Delegate release timer, will call callback function one time
+   * or call delegate every timeout function while stop()
+   * won't be called. It's depend on flag.
    */
-  void SetTimeoutUnsafe(const Milliseconds timeout);
+  class TimerDelegate : public threads::ThreadDelegate {
+   public:
+    /**
+     * @brief Default constructor
+     *
+     * @param timer_thread The Timer_thread pointer
+     */
+    explicit TimerDelegate(Timer* timer);
 
-  /**
-   * @brief startUnsafe, stopUnsafe - methods used for correct synchronization
-   *  and must be used only with sync_primitive (auto_lock e.g.)
-   * @return true if start/stop successfull, false if unsuccessfull
-   */
-  void StartUnsafe();
-  bool StopUnsafe();
+    /**
+     * @brief Destructor
+     */
+    virtual ~TimerDelegate();
 
-  /**
-   * @brief alone function which sends to posix_timer as callee
-   * @param signal_value - structure with parameters of posix_timer callee
-   */
-  friend void HandlePosixTimer(sigval signal_value);
+    /**
+     * @brief Thread main function.
+     */
+    void threadMain() OVERRIDE;
+
+    /**
+     * @brief Called by thread::thread to free all allocated resources.
+     */
+    void exitThreadMain() OVERRIDE;
+
+    /**
+     * @brief Set new Timeout
+     * @param timeout_milliseconds New timeout to be set
+     */
+    void SetTimeOut(const Milliseconds timeout_milliseconds);
+
+    /**
+      * @brief Wait until timer will start
+      */
+    void WaitUntilStart();
+
+    /**
+     * @brief Quits threadMain function after next loop.
+     */
+    void ShouldBeStopped();
+
+    /**
+     * @brief Restarts non-loop timer after current iteration.
+     */
+    void ShouldBeRestarted();
+
+    /**
+     * @brief Return flag IsGoingStop
+     */
+    bool IsGoingToStop() const;
+
+    /**
+     * @brief Gets timeout with overflow check
+     * @return timeout
+     */
+    inline Milliseconds get_timeout() const {
```

Comment:
@AGritsevich, @malirod, As i remember have been decided to avoid header implementation and simplify interfaces (improve readability) by this. If it's possible, please follow this.